File: leet/0063-unique-paths-ii/unique-paths-ii.go
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
    m := len(obstacleGrid)
    n := len(obstacleGrid[0])
    
    cache := make([][]int, m)
    for i := 0; i < m; i++ { cache[i] = make([]int, n) }
    
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if ( i == 0 || j == 0 ) {
                if (obstacleGrid[i][j] == 1) {
                    cache[i][j] = 0
                } else{
                    if (i == 0 && j > 0 && cache[i][j - 1] == 0) || (j == 0 && i > 0 && cache[i-1][j] == 0) {
                        cache[i][j] = 0
                    } else {
                        cache[i][j] = 1
                    }
                }
            } else {
                if (obstacleGrid[i][j] == 1) {
                    cache[i][j] = 0
                } else{
                    cache[i][j] = cache[i - 1][j] + cache[i][j - 1]
                }
            }
        }
    }
    
    return cache[m - 1][n - 1]
}

File: leet/0063-unique-paths-ii/unique-paths-II.js
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
    let m = obstacleGrid.length
    let n = obstacleGrid[0].length

    let mp = new Map();

    let dfs = function(i,j){
        if (i < 0 || j < 0) return 0

        if (obstacleGrid[i][j] == 1) return 0 
        if (i == 0 && j == 0) return 1

        let key = i + "-" + j
        if (mp.has(key)) return mp.get(key)

        let ret = dfs(i - 1, j) + dfs(i, j -1)
        mp.set(key, ret)
        return ret
    }
    
    return dfs(m - 1, n - 1)
};
